<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [ ]:
import pandas as pd
import psycopg2

In [ ]:
# Здесь должны быть параметры для базы данных)

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [20]:
query_3_1 = f'''
select COUNT(*)
from public.vacancies
'''

In [21]:
pd.read_sql_query(query_3_1, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_2784\1467523473.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [8]:
query_3_2 = f'''
select count(*)
from public.employers'''

In [9]:
pd.read_sql_query(query_3_2, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_14776\1321910889.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
query_3_3 = f'''
select count(*)
from public.areas
'''

In [11]:
pd.read_sql_query(query_3_3, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_14776\169222152.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [13]:
query_3_4 = f'''
select count(*)
from public.industries
'''

In [14]:
pd.read_sql_query(query_3_4, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_14776\2934468070.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_4, connection)


count
0    294

***

* Всего 49197 вакансий на 1362 регионов
* Всего 23501 различных работадателей и 294 сфер деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
query_4_1 = f'''
select a.name, count(*)
from public.areas a 
left join public.vacancies v on v.area_id = a.id
group by a.id
order by 2 desc
'''

In [14]:
pd.read_sql_query(query_4_1, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_21880\1702194938.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_1, connection)


name  count
0              Москва   5333
1     Санкт-Петербург   2851
2               Минск   2112
3         Новосибирск   2006
4              Алматы   1892
...               ...    ...
1357            Слуцк      1
1358              Ува      1
1359         Тихорецк      1
1360        Олекминск      1
1361         Павловка      1

[1362 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [23]:
query_4_2 = f'''
select count(*)
from public.vacancies 
where salary_from is not null or salary_to is not null
'''

In [24]:
pd.read_sql_query(query_4_2, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_21880\495373726.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [44]:
query_4_3 = f'''
select
    round(avg(salary_from)) as salary_from,
    round(avg(salary_to)) as salary_to
from public.vacancies
'''

In [45]:
pd.read_sql_query(query_4_3, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_21880\2240822411.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_3, connection)


salary_from  salary_to
0      71065.0   110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [6]:
query_4_4 = '''
select 
    schedule,
    employment, 
    count(*)
from vacancies
group by 1, 2 
order by 3 desc 
'''

In [7]:
pd.read_sql_query(query_4_4, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_14568\3130347181.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [52]:
query_4_5 = '''
select 
    experience, 
    count(*)
from vacancies
group by 1
order by 2
'''

In [53]:
pd.read_sql_query(query_4_5, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22468\3496489661.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

* Большинство вакансий предлагают в городах миллионерах
* У половины вакансий не указана нижняя и верхняя граница заработной платы.
* Самое популярное сочетание рабочего графика и трудоустройства - *'Полный день Полная занятость'*, а далее количество вакансий с другими сочетаниями резко убывает
* Большинство вакансий не требуют большого опыта работы

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [26]:
query_5_1 = '''
(select e.name, count(*)
from vacancies v
join employers e on e.id = v.employer_id
group by e.id
order by 2 desc
limit 1)

union all 

(select e.name, count(*)
from vacancies v
join employers e on e.id = v.employer_id
group by e.id
order by 2 desc
limit 1 offset 4)
'''

In [27]:
pd.read_sql_query(query_5_1, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_2784\2735860424.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
query_5_2 = '''
select
    a.name area,
    count(v.id) vacation,
    count(e.id) company
from areas a
    left join vacancies v on a.id = v.area_id
    left join employers e on a.id = e.area
where v.area_id is null
group by a.name
order by 3 desc
limit 1
'''


In [ ]:
pd.read_sql_query(query_5_2, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22292\1042700087.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_1_2, connection)


area  vacation  company
0  Россия         0      410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [30]:
query_5_3 = '''
select e.name, count(distinct v.area_id)
from vacancies v
join employers e on e.id = v.employer_id
group by e.id
order by 2 desc
'''

In [31]:
pd.read_sql_query(query_5_3, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_2784\3071232566.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_3, connection)


name  count
0                                      Яндекс    181
1                                  Ростелеком    152
2                                  Спецремонт    116
3                      Поляков Денис Иванович     88
4                                   ООО ЕФИН      71
...                                       ...    ...
14901                            НПП Авиатрон      1
14902              Центр дистанционных торгов      1
14903  Городские Телекоммуникационные Системы      1
14904                       Введенский, Отель      1
14905              Группа Компаний Автоимпорт      1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [130]:
query_5_4 = '''
select count(*)
from employers e
left join employers_industries em on em.employer_id = e.id
where industry_id is null
'''

In [131]:
pd.read_sql_query(query_5_4, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22292\2733614102.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [150]:
query_5_5 = '''
select e.name
from employers e 
join employers_industries em on em.employer_id = e.id
group by e.id 
having count(*) = 4
order by 1
offset 2 limit 1
'''

In [151]:
pd.read_sql_query(query_5_5, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22292\3334305160.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [154]:
query_5_6 = '''
select count(*) 
from employers_industries em 
join industries i on i.id = em.industry_id
where i.name = 'Разработка программного обеспечения'
'''

In [155]:
pd.read_sql_query(query_5_6, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22292\2432745783.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [48]:
import requests 
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url) 
page = BeautifulSoup(response.text, 'html.parser')

cities = page.find_all('a')
cities = tuple([city.text for city in cities[41:57]])
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [59]:
query_5_7 = f'''
select a.name, count(v.id)
from vacancies v
join employers e on e.id = v.employer_id
join areas a on a.id = v.area_id
where e.name = 'Яндекс' and a.name in {cities}
group by 1

union 

select 'Total', count(*)
from vacancies v
join employers e on e.id = v.employer_id
join areas a on a.id = v.area_id
where e.name = 'Яндекс' and a.name in {cities}

order by 2
'''

In [60]:
pd.read_sql_query(query_5_7, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_26716\2888435755.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_7, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

### Выводы

* Самое большое количество вакансий предлагает компания *Яндекс*
* Компания *Яндекс* также является лидером по количеству регионов, в котором она публикует свои вакансии
* У 1/3 работадателей не указана сфера деятельности
* У 1/8 работадателей в качестве сферы деятельности указана "Разработка программного обеспечения"
* В городах миллионниках у компании Яндекс примерно одинаковое количество вакансий, однако в городах федерального значения (Санкт-Петербург и Москва) количество вакансий самое наибольшее

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [63]:
query_6_1 = '''
select count(*)
from vacancies
where lower(name) LIKE '%data%' or lower(name) LIKE '%данн%'
'''

In [64]:
pd.read_sql_query(query_6_1, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_2784\2814080296.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
query_6_2 = '''
with t as(
select *
from vacancies
where 
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (lower(name) like '%ml%' and lower(name) not like '%html%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'))
    
select count(*)
from t
where 
    (lower(name) like '%junior%'
    or lower(experience) like '%нет опыта%' 
    or lower(employment) like '%стажировка%')
'''

In [94]:
pd.read_sql_query(query_6_2, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_2784\3416457153.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
query_6_3 = '''
with t as(
select *
from vacancies
where 
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (lower(name) like '%ml%' and lower(name) not like '%html%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'))
    
select count(*)
from t
    where lower(key_skills) like '%sql%'
    or lower(key_skills) like '%postgres%'
'''

In [100]:
pd.read_sql_query(query_6_3, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_2784\133616535.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [101]:
query_6_4 = '''
with t as(
select *
from vacancies
where 
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (lower(name) like '%ml%' and lower(name) not like '%html%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'))
    
select count(*)
from t
    where lower(key_skills) like '%python%'
'''

In [102]:
pd.read_sql_query(query_6_4, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_2784\2825363179.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [7]:
query_6_5 = '''
with t as(
select *
from vacancies
where 
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'))
    
select
    round(avg(length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1), 2)
from t
where key_skills is not null
'''

In [8]:
pd.read_sql_query(query_6_5, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_26716\1879125076.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [17]:
query_6_6 = '''
with t as(
select *
from vacancies
where 
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'))
    
select 
    experience, round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)))
from t 
group by experience

'''

In [18]:
pd.read_sql_query(query_6_6, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_26716\791858229.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_6, connection)


experience     round
0       От 3 до 6 лет  243115.0
1  От 1 года до 3 лет  139675.0
2           Нет опыта   74643.0
3         Более 6 лет       NaN

***

##### Дополнительная таблица, для вывода количества вакансий для дата саентистов!

In [44]:
query_6_7 = '''
select count(*)
from vacancies
where 
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (lower(name) like '%ml%' and lower(name) not like '%html%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
'''

pd.read_sql_query(query_6_7, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_3972\3990311842.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_7, connection)


count
0    536

### Выводы

* 1/50 часть вакансий имеет отношение к данным
* На роль начинающих дата-сайентистов представлена 1/10 часть всех вакансий
* 3/5 часть вакансий требует Python в качестве ключевого навыка, а почти половина вакансий требует знания SQL или postgres
* Чтобы быстрее найти подходящую вакансию, желательно иметь 6 и более ключевых навыков
* С повышением опыта зарплата дата-саентиста увеличивается почти в 2 раза (но почему-то зарплату для опыта более 6 лет предпочитают не указывать)

# Общий вывод по проекту

####  Каких вакансий много? Что можно сказать о рынке труда в различных городах?

* Из запроса, представленного ниже, можно заметить, что больше всего вакансий имеется на такие профессии, как:

    1. Системный администратор (1761)
    2. Программист 1С (1153)
    3. Бизнес-аналитик (488)
    4. Специалист технической поддержки (457)
    5. Системный аналитик (432)

In [30]:
query_7_1 = '''
select name, count(*)
from vacancies
group by name
order by 2 desc
'''

pd.read_sql_query(query_7_1, connection).head(5)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22468\1668112053.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_7_1, connection).head(5)


name  count
0           Системный администратор   1761
1                    Программист 1С   1153
2                   Бизнес-аналитик    488
3  Специалист технической поддержки    457
4                Системный аналитик    432

* В разрезе городов (запрос ниже) можно заметить, что:

    1. В Москве ищут больше всего *бизнес-аналитиков, аналитиков и системных аналитиков* (Скорее всего потому что Москва является самым главным промышленным и экономическим центром России и большинство крупных бизнесов располагаются именно там)
    2. Как в этом, так и в прошлом запросе самой популярной вакансией является *Системный администратор*: спрос на них есть почти в каждом городе, и количество вакансий примерно одинаковое
    3. Спроса на *Программист 1С* больше всего в Новосибирске
    4. *Специалиста технической поддержки* мы не встретили в следующем запросе, хотя он является одной из самых многочисленных вакансий (Может быть потому, что такие специалисты нужны вообще везде, и они равномерно (Хоть и по-немногу) разбросаны по городам)

In [ ]:
query_7_2 = '''
select v.name, a.name, count(*)
from vacancies v
join areas a on a.id = v.area_id
group by 1, 2 
order by 3 desc,2 
'''

pd.read_sql_query(query_7_2, connection).head(10)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22468\4276408171.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_7_1, connection).head(10)


name             name  count
0          Бизнес-аналитик           Москва    173
1                 Аналитик           Москва     69
2  Системный администратор  Санкт-Петербург     68
3       Системный аналитик           Москва     59
4  Системный администратор            Минск     55
5           Программист 1С      Новосибирск     54
6  Системный администратор      Новосибирск     53
7  Системный администратор           Алматы     52
8  Системный администратор     Екатеринбург     51
9  Системный администратор           Москва     50

* Из задания 4.1 Мы узнали, что больше всего вакансий находится в городах: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Этого и следовало ожидать, ведь это огромные промышленные и экономические центры. Но возможно, роль в этом играет еще и население этих городов
* Самые большие средние зарплаты в Германии, Болгарии, Испании, Черногории, Литве. Возможно это связано с тем, что в этих странах цены выше, чем в остальных. А может быть в вакансиях были не заполнены нижние границы зарплатной вилки (СМОТРЕТЬ ЗАПРОС НИЖЕ)


In [48]:
query_7_3 = '''
select a.name, round(avg(coalesce((v.salary_from+v.salary_to)/2, salary_from, salary_to)))
from vacancies v
join areas a on a.id = v.area_id
where v.salary_from is not null or v.salary_to is not null
group by 1
order by 2 desc, 1 
'''
pd.read_sql_query(query_7_3, connection).head(5)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_22468\218697608.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_7_3, connection).head(5)


name     round
0    Германия  452751.0
1    Болгария  324348.0
2     Испания  300000.0
3  Черногория  288589.0
4       Литва  263019.0

* Но из задания 4.3 видно, что в совокупности средние зарплаты уже не такие высокие, как в предыдущем запросе. Это говорит о сильной неравномерности данных. Возможно найдутся даже выбросы
* Почти все вакансии(задание 4.4) рассматривают тип рабочего графика *Полный день - Полная занятость*. Может быть потому, что такой график является наиболее эффективным методом работы
* Больше всего вакансий(задание 4.5) ищут с опытом от 1 года до 3 лет, а меньше всего - более 6 лет. Возможно потому, что на большинство вакансий не требуется очень большой опыт работы. Но при этом вакансий, на которые не требуется опыта, довольно мало, ведь работадатели хотят уже опытных специалистов

####  Какие ещё интересные исследования можно провести?

* Можно узнать, какие навыки чаще всего требуют работадатели у дата саентистов. Исследованя показали, что: 

    1. Самым востребованным навыком является *Python(348)*. Он есть в 3/5 части всех вакансий для дата саентистов. Возможно это потому, что Python является одним из самых главных языков программирования, которые работают с данными
    2. На втором месте располагается *SQL(191)*. Он встречается почти в 2 раза реже, чем Python, но при этом является довольно распостраненным и занимает 2/5 часть всех вакансий. SQL является популярным, удобным и известным языком запросов, поэтому работадатели требуют именно его
    3. Не совсем далеко от SQL ушел и *Machine Learning(114)*. Дата саентистам необходимо разрабатывать свои модели, поэтому работадатели требуют знания в этой области

<u>Всё исследование вы можете рассмотреть ниже</u>

In [ ]:
# Получение всех требуемых навыков для дата саентистов
query_7_4 = '''
with t as(
select *
from vacancies
where 
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'))
    
select
    key_skills
from t
where key_skills is not null
'''

df = pd.read_sql_query(query_7_4, connection)

C:\Users\Aorus\AppData\Local\Temp\ipykernel_3972\3473935026.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_4, connection)


In [40]:
# Словарь для навыков и их количества
skills_dict = {}

# Добавление навыков в словарь
for i in df['key_skills']:
    skills = i.split('\t')
    for j in skills:
        if j in skills_dict:
            skills_dict[j] += 1
        else: 
            skills_dict[j] = 1

# Создание дата фрейма
result = pd.DataFrame( {
    'Название навыка': skills_dict.keys(),
    'Количество вакансий': skills_dict.values()})

# Сортировка по убыванию количества вакансий
result = result.sort_values(by='Количество вакансий', ascending=False)

display(result.head(10))

Название навыка  Количество вакансий
0                      Python                  348
1                         SQL                  191
7            Machine Learning                  114
9                         Git                   66
25  Математическая статистика                   62
40              Data Analysis                   54
45               Data Science                   52
63                     Pandas                   52
4                          ML                   49
43            Английский язык                   48